src/ripple/protocol/impl/STTx.cpp

```diff
@@ -307,13 +307,63 @@ isMemoOkay (STObject const& st, std::string& reason)
 
         for (auto const& memoElement : *memoObj)
         {
-            if ((memoElement.getFName() != sfMemoType) &&
-                (memoElement.getFName() != sfMemoData) &&
-                (memoElement.getFName() != sfMemoFormat))
+            auto const& name = memoElement.getFName();
+
+            if (name != sfMemoType &&
+                name != sfMemoData &&
+                name != sfMemoFormat)
+            {
+                reason = "A memo may contain only MemoType, MemoData or "
+                         "MemoFormat fields.";
+                return false;
+            }
+
+            // The raw data is stored as hex-octets, which we want to decode.
+            auto data = strUnHex (memoElement.getText ());
+
+            if (!data.second)
             {
-                reason = "A memo may contain only MemoType, MemoData or MemoFormat fields.";
+                reason = "The MemoType, MemoData and MemoFormat fields may "
+                         "only contain hex-encoded data.";
                 return false;
             }
+
+            if (name == sfMemoData)
+                continue;
+
+            // The only allowed characters for MemoType and MemoFormat are the
+            // characters allowed in URLs per RFC 3986: alphanumerics and the
+            // following symbols: -._~:/?#[]@!$&'()*+,;=%
+            static int const allowed_symbols[] =
```

Comment:
Consider using `bool` instead.  In general one of the most expensive things we do on a computer is a data cache miss.  So smaller things often execute faster, even if it takes extra instructions to work with it.  I'll stop short of suggesting a `std::bitset` though as that may be too extreme.

This is just a suggestion.  Nothing I'm loosing sleep over.